// @ag-ui/core/dist/index.js
let zod = require("zod");

//#region src/types.ts
const FunctionCallSchema = zod.z.object({
	name: zod.z.string(),
	arguments: zod.z.string()
});
const ToolCallSchema = zod.z.object({
	id: zod.z.string(),
	type: zod.z.literal("function"),
	function: FunctionCallSchema,
	encryptedValue: zod.z.string().optional()
});
const BaseMessageSchema = zod.z.object({
	id: zod.z.string(),
	role: zod.z.string(),
	content: zod.z.string().optional(),
	name: zod.z.string().optional(),
	encryptedValue: zod.z.string().optional()
});
const TextInputContentSchema = zod.z.object({
	type: zod.z.literal("text"),
	text: zod.z.string()
});
const InputContentDataSourceSchema = zod.z.object({
	type: zod.z.literal("data"),
	value: zod.z.string(),
	mimeType: zod.z.string()
});
const InputContentUrlSourceSchema = zod.z.object({
	type: zod.z.literal("url"),
	value: zod.z.string(),
	mimeType: zod.z.string().optional()
});
const InputContentSourceSchema = zod.z.discriminatedUnion("type", [InputContentDataSourceSchema, InputContentUrlSourceSchema]);
const ImageInputContentSchema = zod.z.object({
	type: zod.z.literal("image"),
	source: InputContentSourceSchema,
	metadata: zod.z.unknown().optional()
});
const AudioInputContentSchema = zod.z.object({
	type: zod.z.literal("audio"),
	source: InputContentSourceSchema,
	metadata: zod.z.unknown().optional()
});
const VideoInputContentSchema = zod.z.object({
	type: zod.z.literal("video"),
	source: InputContentSourceSchema,
	metadata: zod.z.unknown().optional()
});
const DocumentInputContentSchema = zod.z.object({
	type: zod.z.literal("document"),
	source: InputContentSourceSchema,
	metadata: zod.z.unknown().optional()
});
const ImageInputPartSchema = ImageInputContentSchema;
const AudioInputPartSchema = AudioInputContentSchema;
const VideoInputPartSchema = VideoInputContentSchema;
const DocumentInputPartSchema = DocumentInputContentSchema;
const LegacyBinaryInputContentObjectSchema = zod.z.object({
	type: zod.z.literal("binary"),
	mimeType: zod.z.string(),
	id: zod.z.string().optional(),
	url: zod.z.string().optional(),
	data: zod.z.string().optional(),
	filename: zod.z.string().optional()
});
const ensureBinaryPayload = (value, ctx) => {
	if (!value.id && !value.url && !value.data) ctx.addIssue({
		code: zod.z.ZodIssueCode.custom,
		message: "BinaryInputContent requires at least one of id, url, or data.",
		path: ["id"]
	});
};
const BinaryInputContentSchema = LegacyBinaryInputContentObjectSchema.superRefine((value, ctx) => {
	ensureBinaryPayload(value, ctx);
});
const InputContentBaseSchema = zod.z.discriminatedUnion("type", [
	TextInputContentSchema,
	ImageInputContentSchema,
	AudioInputContentSchema,
	VideoInputContentSchema,
	DocumentInputContentSchema,
	LegacyBinaryInputContentObjectSchema
]);
const InputContentSchema = InputContentBaseSchema.superRefine((value, ctx) => {
	if (value.type === "binary") ensureBinaryPayload(value, ctx);
});
const DeveloperMessageSchema = BaseMessageSchema.extend({
	role: zod.z.literal("developer"),
	content: zod.z.string()
});
const SystemMessageSchema = BaseMessageSchema.extend({
	role: zod.z.literal("system"),
	content: zod.z.string()
});
const AssistantMessageSchema = BaseMessageSchema.extend({
	role: zod.z.literal("assistant"),
	content: zod.z.string().optional(),
	toolCalls: zod.z.array(ToolCallSchema).optional()
});
const UserMessageSchema = BaseMessageSchema.extend({
	role: zod.z.literal("user"),
	content: zod.z.union([zod.z.string(), zod.z.array(InputContentSchema)])
});
const ToolMessageSchema = zod.z.object({
	id: zod.z.string(),
	content: zod.z.string(),
	role: zod.z.literal("tool"),
	toolCallId: zod.z.string(),
	error: zod.z.string().optional(),
	encryptedValue: zod.z.string().optional()
});
const ActivityMessageSchema = zod.z.object({
	id: zod.z.string(),
	role: zod.z.literal("activity"),
	activityType: zod.z.string(),
	content: zod.z.record(zod.z.any())
});
const ReasoningMessageSchema = zod.z.object({
	id: zod.z.string(),
	role: zod.z.literal("reasoning"),
	content: zod.z.string(),
	encryptedValue: zod.z.string().optional()
});
const MessageSchema = zod.z.discriminatedUnion("role", [
	DeveloperMessageSchema,
	SystemMessageSchema,
	AssistantMessageSchema,
	UserMessageSchema,
	ToolMessageSchema,
	ActivityMessageSchema,
	ReasoningMessageSchema
]);
const RoleSchema = zod.z.union([
	zod.z.literal("developer"),
	zod.z.literal("system"),
	zod.z.literal("assistant"),
	zod.z.literal("user"),
	zod.z.literal("tool"),
	zod.z.literal("activity"),
	zod.z.literal("reasoning")
]);
const ContextSchema = zod.z.object({
	description: zod.z.string(),
	value: zod.z.string()
});
const ToolSchema = zod.z.object({
	name: zod.z.string(),
	description: zod.z.string(),
	parameters: zod.z.any(),
	metadata: zod.z.record(zod.z.any()).optional()
});
const InterruptSchema = zod.z.object({
	id: zod.z.string(),
	reason: zod.z.string(),
	message: zod.z.string().optional(),
	toolCallId: zod.z.string().optional(),
	responseSchema: zod.z.record(zod.z.any()).optional(),
	expiresAt: zod.z.string().optional(),
	metadata: zod.z.record(zod.z.any()).optional()
});
const ResumeEntrySchema = zod.z.object({
	interruptId: zod.z.string(),
	status: zod.z.enum(["resolved", "cancelled"]),
	payload: zod.z.any().optional()
});
const RunAgentInputSchema = zod.z.object({
	threadId: zod.z.string(),
	runId: zod.z.string(),
	parentRunId: zod.z.string().optional(),
	state: zod.z.any(),
	messages: zod.z.array(MessageSchema),
	tools: zod.z.array(ToolSchema),
	context: zod.z.array(ContextSchema),
	forwardedProps: zod.z.any(),
	resume: zod.z.array(ResumeEntrySchema).optional()
});
const StateSchema = zod.z.any();
var AGUIError = class extends Error {
	constructor(message) {
		super(message);
	}
};
var AGUIConnectNotImplementedError = class extends AGUIError {
	constructor() {
		super("Connect not implemented. This method is not supported by the current agent.");
	}
};

//#endregion
//#region src/capabilities.ts
/** Describes a sub-agent that can be invoked by a parent agent. */
const SubAgentInfoSchema = zod.z.object({
	name: zod.z.string(),
	description: zod.z.string().optional()
});
/**
* Basic metadata about the agent. Useful for discovery UIs, agent marketplaces,
* and debugging. Set these when you want clients to display agent information
* or when multiple agents are available and users need to pick one.
*/
const IdentityCapabilitiesSchema = zod.z.object({
	name: zod.z.string().optional(),
	type: zod.z.string().optional(),
	description: zod.z.string().optional(),
	version: zod.z.string().optional(),
	provider: zod.z.string().optional(),
	documentationUrl: zod.z.string().optional(),
	metadata: zod.z.record(zod.z.unknown()).optional()
});
/**
* Declares which transport mechanisms the agent supports. Clients use this
* to pick the best connection strategy. Only set flags to `true` for transports
* your agent actually handles — omit or set `false` for unsupported ones.
*/
const TransportCapabilitiesSchema = zod.z.object({
	streaming: zod.z.boolean().optional(),
	websocket: zod.z.boolean().optional(),
	httpBinary: zod.z.boolean().optional(),
	pushNotifications: zod.z.boolean().optional(),
	resumable: zod.z.boolean().optional()
});
/**
* Tool calling capabilities. Distinguishes between tools the agent itself provides
* (listed in `items`) and tools the client passes at runtime via `RunAgentInput.tools`.
* Enable this when your agent can call functions, search the web, execute code, etc.
*/
const ToolsCapabilitiesSchema = zod.z.object({
	supported: zod.z.boolean().optional(),
	items: zod.z.array(ToolSchema).optional(),
	parallelCalls: zod.z.boolean().optional(),
	clientProvided: zod.z.boolean().optional()
});
/**
* Output format support. Enable `structuredOutput` when your agent can return
* responses conforming to a JSON schema, which is useful for programmatic consumption.
*/
const OutputCapabilitiesSchema = zod.z.object({
	structuredOutput: zod.z.boolean().optional(),
	supportedMimeTypes: zod.z.array(zod.z.string()).optional()
});
/**
* State and memory management capabilities. These tell the client how the agent
* handles shared state and whether conversation context persists across runs.
*/
const StateCapabilitiesSchema = zod.z.object({
	snapshots: zod.z.boolean().optional(),
	deltas: zod.z.boolean().optional(),
	memory: zod.z.boolean().optional(),
	persistentState: zod.z.boolean().optional()
});
/**
* Multi-agent coordination capabilities. Enable these when your agent can
* orchestrate or hand off work to other agents.
*/
const MultiAgentCapabilitiesSchema = zod.z.object({
	supported: zod.z.boolean().optional(),
	delegation: zod.z.boolean().optional(),
	handoffs: zod.z.boolean().optional(),
	subAgents: zod.z.array(SubAgentInfoSchema).optional()
});
/**
* Reasoning and thinking capabilities. Enable these when your agent exposes its
* internal thought process (e.g., chain-of-thought, extended thinking).
*/
const ReasoningCapabilitiesSchema = zod.z.object({
	supported: zod.z.boolean().optional(),
	streaming: zod.z.boolean().optional(),
	encrypted: zod.z.boolean().optional()
});
/**
* Modalities the agent can accept as input. Clients use this to show/hide
* file upload buttons, audio recorders, image pickers, etc.
*/
const MultimodalInputCapabilitiesSchema = zod.z.object({
	image: zod.z.boolean().optional(),
	audio: zod.z.boolean().optional(),
	video: zod.z.boolean().optional(),
	pdf: zod.z.boolean().optional(),
	file: zod.z.boolean().optional()
});
/**
* Modalities the agent can produce as output. Clients use this to anticipate
* rich content in the agent's response.
*/
const MultimodalOutputCapabilitiesSchema = zod.z.object({
	image: zod.z.boolean().optional(),
	audio: zod.z.boolean().optional()
});
/**
* Multimodal input and output support. Organized into `input` and `output`
* sub-objects so clients can independently query what the agent accepts
* versus what it produces.
*/
const MultimodalCapabilitiesSchema = zod.z.object({
	input: MultimodalInputCapabilitiesSchema.optional(),
	output: MultimodalOutputCapabilitiesSchema.optional()
});
/**
* Execution control and limits. Declare these so clients can set expectations
* about how long or how many steps an agent run might take.
*/
const ExecutionCapabilitiesSchema = zod.z.object({
	codeExecution: zod.z.boolean().optional(),
	sandboxed: zod.z.boolean().optional(),
	maxIterations: zod.z.number().optional(),
	maxExecutionTime: zod.z.number().optional()
});
/**
* Human-in-the-loop interaction support. Enable these when your agent can pause
* execution to request human input, approval, or feedback before continuing.
*/
const HumanInTheLoopCapabilitiesSchema = zod.z.object({
	supported: zod.z.boolean().optional(),
	approvals: zod.z.boolean().optional(),
	interventions: zod.z.boolean().optional(),
	feedback: zod.z.boolean().optional(),
	interrupts: zod.z.boolean().optional(),
	approveWithEdits: zod.z.boolean().optional()
});
/**
* A typed, categorized snapshot of an agent's current capabilities.
* Returned by `getCapabilities()` on `AbstractAgent`.
*
* All fields are optional — agents only declare what they support.
* Omitted fields mean the capability is not declared (unknown), not that
* it's unsupported.
*
* The `custom` field is an escape hatch for integration-specific capabilities
* that don't fit into the standard categories.
*/
const AgentCapabilitiesSchema = zod.z.object({
	identity: IdentityCapabilitiesSchema.optional(),
	transport: TransportCapabilitiesSchema.optional(),
	tools: ToolsCapabilitiesSchema.optional(),
	output: OutputCapabilitiesSchema.optional(),
	state: StateCapabilitiesSchema.optional(),
	multiAgent: MultiAgentCapabilitiesSchema.optional(),
	reasoning: ReasoningCapabilitiesSchema.optional(),
	multimodal: MultimodalCapabilitiesSchema.optional(),
	execution: ExecutionCapabilitiesSchema.optional(),
	humanInTheLoop: HumanInTheLoopCapabilitiesSchema.optional(),
	custom: zod.z.record(zod.z.unknown()).optional()
});

//#endregion
//#region src/events.ts
const TextMessageRoleSchema = zod.z.union([
	zod.z.literal("developer"),
	zod.z.literal("system"),
	zod.z.literal("assistant"),
	zod.z.literal("user")
]);
let EventType = /* @__PURE__ */ function(EventType) {
	EventType["TEXT_MESSAGE_START"] = "TEXT_MESSAGE_START";
	EventType["TEXT_MESSAGE_CONTENT"] = "TEXT_MESSAGE_CONTENT";
	EventType["TEXT_MESSAGE_END"] = "TEXT_MESSAGE_END";
	EventType["TEXT_MESSAGE_CHUNK"] = "TEXT_MESSAGE_CHUNK";
	EventType["TOOL_CALL_START"] = "TOOL_CALL_START";
	EventType["TOOL_CALL_ARGS"] = "TOOL_CALL_ARGS";
	EventType["TOOL_CALL_END"] = "TOOL_CALL_END";
	EventType["TOOL_CALL_CHUNK"] = "TOOL_CALL_CHUNK";
	EventType["TOOL_CALL_RESULT"] = "TOOL_CALL_RESULT";
	/**
	* @deprecated Use REASONING_START instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_START"] = "THINKING_START";
	/**
	* @deprecated Use REASONING_END instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_END"] = "THINKING_END";
	/**
	* @deprecated Use REASONING_MESSAGE_START instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_START"] = "THINKING_TEXT_MESSAGE_START";
	/**
	* @deprecated Use REASONING_MESSAGE_CONTENT instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_CONTENT"] = "THINKING_TEXT_MESSAGE_CONTENT";
	/**
	* @deprecated Use REASONING_MESSAGE_END instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_END"] = "THINKING_TEXT_MESSAGE_END";
	EventType["STATE_SNAPSHOT"] = "STATE_SNAPSHOT";
	EventType["STATE_DELTA"] = "STATE_DELTA";
	EventType["MESSAGES_SNAPSHOT"] = "MESSAGES_SNAPSHOT";
	EventType["ACTIVITY_SNAPSHOT"] = "ACTIVITY_SNAPSHOT";
	EventType["ACTIVITY_DELTA"] = "ACTIVITY_DELTA";
	EventType["RAW"] = "RAW";
	EventType["CUSTOM"] = "CUSTOM";
	EventType["RUN_STARTED"] = "RUN_STARTED";
	EventType["RUN_FINISHED"] = "RUN_FINISHED";
	EventType["RUN_ERROR"] = "RUN_ERROR";
	EventType["STEP_STARTED"] = "STEP_STARTED";
	EventType["STEP_FINISHED"] = "STEP_FINISHED";
	EventType["REASONING_START"] = "REASONING_START";
	EventType["REASONING_MESSAGE_START"] = "REASONING_MESSAGE_START";
	EventType["REASONING_MESSAGE_CONTENT"] = "REASONING_MESSAGE_CONTENT";
	EventType["REASONING_MESSAGE_END"] = "REASONING_MESSAGE_END";
	EventType["REASONING_MESSAGE_CHUNK"] = "REASONING_MESSAGE_CHUNK";
	EventType["REASONING_END"] = "REASONING_END";
	EventType["REASONING_ENCRYPTED_VALUE"] = "REASONING_ENCRYPTED_VALUE";
	return EventType;
}({});
const BaseEventSchema = zod.z.object({
	type: zod.z.nativeEnum(EventType),
	timestamp: zod.z.number().optional(),
	rawEvent: zod.z.any().optional()
}).passthrough();
const TextMessageStartEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TEXT_MESSAGE_START),
	messageId: zod.z.string(),
	role: TextMessageRoleSchema.default("assistant"),
	name: zod.z.string().optional()
});
const TextMessageContentEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TEXT_MESSAGE_CONTENT),
	messageId: zod.z.string(),
	delta: zod.z.string()
});
const TextMessageEndEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TEXT_MESSAGE_END),
	messageId: zod.z.string()
});
const TextMessageChunkEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TEXT_MESSAGE_CHUNK),
	messageId: zod.z.string().optional(),
	role: TextMessageRoleSchema.optional(),
	delta: zod.z.string().optional(),
	name: zod.z.string().optional()
});
/**
* @deprecated Use ReasoningMessageStartEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageStartEventSchema = BaseEventSchema.extend({ type: zod.z.literal(EventType.THINKING_TEXT_MESSAGE_START) });
/**
* @deprecated Use ReasoningMessageContentEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageContentEventSchema = TextMessageContentEventSchema.omit({
	messageId: true,
	type: true
}).extend({ type: zod.z.literal(EventType.THINKING_TEXT_MESSAGE_CONTENT) });
/**
* @deprecated Use ReasoningMessageEndEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageEndEventSchema = BaseEventSchema.extend({ type: zod.z.literal(EventType.THINKING_TEXT_MESSAGE_END) });
const ToolCallStartEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TOOL_CALL_START),
	toolCallId: zod.z.string(),
	toolCallName: zod.z.string(),
	parentMessageId: zod.z.string().optional()
});
const ToolCallArgsEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TOOL_CALL_ARGS),
	toolCallId: zod.z.string(),
	delta: zod.z.string()
});
const ToolCallEndEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TOOL_CALL_END),
	toolCallId: zod.z.string()
});
const ToolCallResultEventSchema = BaseEventSchema.extend({
	messageId: zod.z.string(),
	type: zod.z.literal(EventType.TOOL_CALL_RESULT),
	toolCallId: zod.z.string(),
	content: zod.z.string(),
	role: zod.z.literal("tool").optional()
});
const ToolCallChunkEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.TOOL_CALL_CHUNK),
	toolCallId: zod.z.string().optional(),
	toolCallName: zod.z.string().optional(),
	parentMessageId: zod.z.string().optional(),
	delta: zod.z.string().optional()
});
/**
* @deprecated Use ReasoningStartEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingStartEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.THINKING_START),
	title: zod.z.string().optional()
});
/**
* @deprecated Use ReasoningEndEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingEndEventSchema = BaseEventSchema.extend({ type: zod.z.literal(EventType.THINKING_END) });
const StateSnapshotEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.STATE_SNAPSHOT),
	snapshot: StateSchema
});
const StateDeltaEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.STATE_DELTA),
	delta: zod.z.array(zod.z.any())
});
const MessagesSnapshotEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.MESSAGES_SNAPSHOT),
	messages: zod.z.array(MessageSchema)
});
const ActivitySnapshotEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.ACTIVITY_SNAPSHOT),
	messageId: zod.z.string(),
	activityType: zod.z.string(),
	content: zod.z.record(zod.z.any()),
	replace: zod.z.boolean().optional().default(true)
});
const ActivityDeltaEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.ACTIVITY_DELTA),
	messageId: zod.z.string(),
	activityType: zod.z.string(),
	patch: zod.z.array(zod.z.any())
});
const RawEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.RAW),
	event: zod.z.any(),
	source: zod.z.string().optional()
});
const CustomEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.CUSTOM),
	name: zod.z.string(),
	value: zod.z.any()
});
const RunStartedEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.RUN_STARTED),
	threadId: zod.z.string(),
	runId: zod.z.string(),
	parentRunId: zod.z.string().optional(),
	input: RunAgentInputSchema.optional()
});
const RunFinishedSuccessOutcomeSchema = zod.z.object({ type: zod.z.literal("success") }).strict();
const RunFinishedInterruptOutcomeSchema = zod.z.object({
	type: zod.z.literal("interrupt"),
	interrupts: zod.z.array(InterruptSchema).min(1)
}).strict();
const RunFinishedOutcomeSchema = zod.z.discriminatedUnion("type", [RunFinishedSuccessOutcomeSchema, RunFinishedInterruptOutcomeSchema]);
const RunFinishedEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.RUN_FINISHED),
	threadId: zod.z.string(),
	runId: zod.z.string(),
	result: zod.z.any().optional(),
	outcome: RunFinishedOutcomeSchema.nullable().optional().transform((v) => v ?? void 0)
});
const RunErrorEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.RUN_ERROR),
	message: zod.z.string(),
	code: zod.z.string().optional()
});
const StepStartedEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.STEP_STARTED),
	stepName: zod.z.string()
});
const StepFinishedEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.STEP_FINISHED),
	stepName: zod.z.string()
});
const ReasoningEncryptedValueSubtypeSchema = zod.z.union([zod.z.literal("tool-call"), zod.z.literal("message")]);
const ReasoningStartEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_START),
	messageId: zod.z.string()
});
const ReasoningMessageStartEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_MESSAGE_START),
	messageId: zod.z.string(),
	role: zod.z.literal("reasoning")
});
const ReasoningMessageContentEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_MESSAGE_CONTENT),
	messageId: zod.z.string(),
	delta: zod.z.string()
});
const ReasoningMessageEndEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_MESSAGE_END),
	messageId: zod.z.string()
});
const ReasoningMessageChunkEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_MESSAGE_CHUNK),
	messageId: zod.z.string().optional(),
	delta: zod.z.string().optional()
});
const ReasoningEndEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_END),
	messageId: zod.z.string()
});
const ReasoningEncryptedValueEventSchema = BaseEventSchema.extend({
	type: zod.z.literal(EventType.REASONING_ENCRYPTED_VALUE),
	subtype: ReasoningEncryptedValueSubtypeSchema,
	entityId: zod.z.string(),
	encryptedValue: zod.z.string()
});
const EventSchemas = zod.z.discriminatedUnion("type", [
	TextMessageStartEventSchema,
	TextMessageContentEventSchema,
	TextMessageEndEventSchema,
	TextMessageChunkEventSchema,
	ThinkingStartEventSchema,
	ThinkingEndEventSchema,
	ThinkingTextMessageStartEventSchema,
	ThinkingTextMessageContentEventSchema,
	ThinkingTextMessageEndEventSchema,
	ToolCallStartEventSchema,
	ToolCallArgsEventSchema,
	ToolCallEndEventSchema,
	ToolCallChunkEventSchema,
	ToolCallResultEventSchema,
	StateSnapshotEventSchema,
	StateDeltaEventSchema,
	MessagesSnapshotEventSchema,
	ActivitySnapshotEventSchema,
	ActivityDeltaEventSchema,
	RawEventSchema,
	CustomEventSchema,
	RunStartedEventSchema,
	RunFinishedEventSchema,
	RunErrorEventSchema,
	StepStartedEventSchema,
	StepFinishedEventSchema,
	ReasoningStartEventSchema,
	ReasoningMessageStartEventSchema,
	ReasoningMessageContentEventSchema,
	ReasoningMessageEndEventSchema,
	ReasoningMessageChunkEventSchema,
	ReasoningEndEventSchema,
	ReasoningEncryptedValueEventSchema
]);

//#endregion
exports.AGUIConnectNotImplementedError = AGUIConnectNotImplementedError;
exports.AGUIError = AGUIError;
exports.ActivityDeltaEventSchema = ActivityDeltaEventSchema;
exports.ActivityMessageSchema = ActivityMessageSchema;
exports.ActivitySnapshotEventSchema = ActivitySnapshotEventSchema;
exports.AgentCapabilitiesSchema = AgentCapabilitiesSchema;
exports.AssistantMessageSchema = AssistantMessageSchema;
exports.AudioInputContentSchema = AudioInputContentSchema;
exports.AudioInputPartSchema = AudioInputPartSchema;
exports.BaseEventSchema = BaseEventSchema;
exports.BaseMessageSchema = BaseMessageSchema;
exports.BinaryInputContentSchema = BinaryInputContentSchema;
exports.ContextSchema = ContextSchema;
exports.CustomEventSchema = CustomEventSchema;
exports.DeveloperMessageSchema = DeveloperMessageSchema;
exports.DocumentInputContentSchema = DocumentInputContentSchema;
exports.DocumentInputPartSchema = DocumentInputPartSchema;
exports.EventSchemas = EventSchemas;
exports.EventType = EventType;
exports.ExecutionCapabilitiesSchema = ExecutionCapabilitiesSchema;
exports.FunctionCallSchema = FunctionCallSchema;
exports.HumanInTheLoopCapabilitiesSchema = HumanInTheLoopCapabilitiesSchema;
exports.IdentityCapabilitiesSchema = IdentityCapabilitiesSchema;
exports.ImageInputContentSchema = ImageInputContentSchema;
exports.ImageInputPartSchema = ImageInputPartSchema;
exports.InputContentDataSourceSchema = InputContentDataSourceSchema;
exports.InputContentSchema = InputContentSchema;
exports.InputContentSourceSchema = InputContentSourceSchema;
exports.InputContentUrlSourceSchema = InputContentUrlSourceSchema;
exports.InterruptSchema = InterruptSchema;
exports.MessageSchema = MessageSchema;
exports.MessagesSnapshotEventSchema = MessagesSnapshotEventSchema;
exports.MultiAgentCapabilitiesSchema = MultiAgentCapabilitiesSchema;
exports.MultimodalCapabilitiesSchema = MultimodalCapabilitiesSchema;
exports.MultimodalInputCapabilitiesSchema = MultimodalInputCapabilitiesSchema;
exports.MultimodalOutputCapabilitiesSchema = MultimodalOutputCapabilitiesSchema;
exports.OutputCapabilitiesSchema = OutputCapabilitiesSchema;
exports.RawEventSchema = RawEventSchema;
exports.ReasoningCapabilitiesSchema = ReasoningCapabilitiesSchema;
exports.ReasoningEncryptedValueEventSchema = ReasoningEncryptedValueEventSchema;
exports.ReasoningEncryptedValueSubtypeSchema = ReasoningEncryptedValueSubtypeSchema;
exports.ReasoningEndEventSchema = ReasoningEndEventSchema;
exports.ReasoningMessageChunkEventSchema = ReasoningMessageChunkEventSchema;
exports.ReasoningMessageContentEventSchema = ReasoningMessageContentEventSchema;
exports.ReasoningMessageEndEventSchema = ReasoningMessageEndEventSchema;
exports.ReasoningMessageSchema = ReasoningMessageSchema;
exports.ReasoningMessageStartEventSchema = ReasoningMessageStartEventSchema;
exports.ReasoningStartEventSchema = ReasoningStartEventSchema;
exports.ResumeEntrySchema = ResumeEntrySchema;
exports.RoleSchema = RoleSchema;
exports.RunAgentInputSchema = RunAgentInputSchema;
exports.RunErrorEventSchema = RunErrorEventSchema;
exports.RunFinishedEventSchema = RunFinishedEventSchema;
exports.RunFinishedInterruptOutcomeSchema = RunFinishedInterruptOutcomeSchema;
exports.RunFinishedOutcomeSchema = RunFinishedOutcomeSchema;
exports.RunFinishedSuccessOutcomeSchema = RunFinishedSuccessOutcomeSchema;
exports.RunStartedEventSchema = RunStartedEventSchema;
exports.StateCapabilitiesSchema = StateCapabilitiesSchema;
exports.StateDeltaEventSchema = StateDeltaEventSchema;
exports.StateSchema = StateSchema;
exports.StateSnapshotEventSchema = StateSnapshotEventSchema;
exports.StepFinishedEventSchema = StepFinishedEventSchema;
exports.StepStartedEventSchema = StepStartedEventSchema;
exports.SubAgentInfoSchema = SubAgentInfoSchema;
exports.SystemMessageSchema = SystemMessageSchema;
exports.TextInputContentSchema = TextInputContentSchema;
exports.TextMessageChunkEventSchema = TextMessageChunkEventSchema;
exports.TextMessageContentEventSchema = TextMessageContentEventSchema;
exports.TextMessageEndEventSchema = TextMessageEndEventSchema;
exports.TextMessageStartEventSchema = TextMessageStartEventSchema;
exports.ThinkingEndEventSchema = ThinkingEndEventSchema;
exports.ThinkingStartEventSchema = ThinkingStartEventSchema;
exports.ThinkingTextMessageContentEventSchema = ThinkingTextMessageContentEventSchema;
exports.ThinkingTextMessageEndEventSchema = ThinkingTextMessageEndEventSchema;
exports.ThinkingTextMessageStartEventSchema = ThinkingTextMessageStartEventSchema;
exports.ToolCallArgsEventSchema = ToolCallArgsEventSchema;
exports.ToolCallChunkEventSchema = ToolCallChunkEventSchema;
exports.ToolCallEndEventSchema = ToolCallEndEventSchema;
exports.ToolCallResultEventSchema = ToolCallResultEventSchema;
exports.ToolCallSchema = ToolCallSchema;
exports.ToolCallStartEventSchema = ToolCallStartEventSchema;
exports.ToolMessageSchema = ToolMessageSchema;
exports.ToolSchema = ToolSchema;
exports.ToolsCapabilitiesSchema = ToolsCapabilitiesSchema;
exports.TransportCapabilitiesSchema = TransportCapabilitiesSchema;
exports.UserMessageSchema = UserMessageSchema;
exports.VideoInputContentSchema = VideoInputContentSchema;
exports.VideoInputPartSchema = VideoInputPartSchema;
